src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubAsyncCollector.cs

```diff
@@ -1,50 +1,93 @@
 
+using System;
 using System.Collections.Generic;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.ServiceBus.Messaging;
 
 namespace Microsoft.Azure.WebJobs.ServiceBus
 {
-    // Core object to send events to EventHub. 
-    // Any user parameter that sends events will eventually get bound to this object. 
-    // This gets wrappers with various adapters and passed to the user function. 
-    internal class EventHubAsyncCollector : IAsyncCollector<EventData>
+    /// <summary>
+    /// Core object to send events to EventHub. 
+    /// Any user parameter that sends EventHub events will eventually get bound to this object. 
+    /// This will queue events and send in batches, also keeping under the 256kb event hub limit per batch. 
+    /// </summary>
+    public class EventHubAsyncCollector : IAsyncCollector<EventData>
     {
         private readonly EventHubClient _client;
 
         private List<EventData> _list = new List<EventData>();
+
+        // total size of bytes in _list that we'll be sending in this batch. 
+        private int _currentByteSize = 0;
+
         private const int BatchSize = 100;
 
+        // Suggested to use 240k instead of 256k to leave padding room for headers.
+        private const int MaxByteSize = 240 * 1000; 
+
+        /// <summary>
+        /// Create a sender around the given client. 
+        /// </summary>
+        /// <param name="client"></param>
         public EventHubAsyncCollector(EventHubClient client)
         {
             _client = client;
         }
 
-        public async Task AddAsync(EventData eventData, CancellationToken cancellationToken = default(CancellationToken))
+        /// <summary>
+        /// Add an event. 
+        /// </summary>
+        /// <param name="item">The event to add</param>
+        /// <param name="cancellationToken">a cancellation token. </param>
+        /// <returns></returns>
+        public async Task AddAsync(EventData item, CancellationToken cancellationToken = default(CancellationToken))
         {
-            bool flush;
-            lock (_list)
+            while (true)
             {
-                _list.Add(eventData);
-                flush = _list.Count > BatchSize;
-            }
+                lock (_list)
+                {
+                    var size = (int)item.SerializedSizeInBytes;
 
-            if (flush)
-            {
-                await this.FlushAsync(cancellationToken);
+                    if (size > MaxByteSize)
+                    {
+                        // Single event is too large to add.
+                        string msg = string.Format("Event is too large. Event is approximately {0}b and max size is {0}b", size, MaxByteSize);
+                        throw new InvalidOperationException(msg);
+                    }
+
+                    if ((_currentByteSize + size > MaxByteSize) || (_list.Count >= BatchSize))
```

Comment:
Yes.  I thought it'd be easier to follow the logic like this, but I can flip it.